File: csharp/demo/demo21a03-condition-variable.cs
/*
 * CONDITION VARIABLES
 */
using System;
using System.Threading;



class Demo21A03 : IRunnable
{
    public void run()
    {
        var conditionVar = new object();

        ThreadStart foo = () =>
        {
            Console.WriteLine("foo is waiting...");

            lock (conditionVar)
            {
                Monitor.Wait(conditionVar);
            }

            Console.WriteLine("foo resumed");
        };

        ThreadStart bar = () =>
        {
            Thread.Sleep(3000);

            lock (conditionVar)
            {
                // Notify all waiting threads
                Monitor.PulseAll(conditionVar);
            }
        };

        for (int i = 0; i < 3; ++i)
            new Thread(foo).Start();

        new Thread(bar).Start();
    }
}

File: python/demo21a03_condition_variable.py
'''
CONDITION VARIABLES
'''

import time
import threading



condition_var = threading.Condition()



def foo():
    print('foo is waiting...')
    with condition_var:
        condition_var.wait()
    print('foo resumed')



def bar():
    time.sleep(3)
    with condition_var:
        # Notify all waiting threads
        condition_var.notify_all()



_ = [ threading.Thread(target=foo).start() for _ in range(3) ]
threading.Thread(target=bar).start()
